contracts/plugins/trading/DutchTradeRouter.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: BlueOak-1.0.0
+pragma solidity 0.8.19;
+
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IDutchTradeCallee, TradeStatus, DutchTrade } from "../trading/DutchTrade.sol";
+import { IMain } from "../../interfaces/IMain.sol";
+
+contract DutchTradeRouter is IDutchTradeCallee {
+    using SafeERC20 for IERC20;
+    struct Bid {
+        /// @notice The DutchTrade that was bid on
+        DutchTrade trade;
+        /// @notice The token sold to the protocol
+        IERC20 sellToken;
+        /// @notice The amount of tokenIn the protocol got {qSellAmt}
+        uint256 sellAmt;
+        /// @notice The token bought from the trade
+        IERC20 buyToken;
+        /// @notice The amount of tokenOut the we got {qBuyAmt}
+        uint256 buyAmt;
+    }
+
+    /// @notice Emitted when a bid is placed
+    /// @param main The main contract of the rToken
+    /// @param trade The DutchTrade that was bid on
+    /// @param bidder The address of the bidder
+    /// @param sellToken The token being sold by the protocol
+    /// @param soldAmt The amount of sellToken sold {qSellToken}
+    /// @param buyToken The token being bought by the protocol
+    /// @param boughtAmt The amount of buyToken bought {qBuyToken}
+    event BidPlaced(
+        IMain main,
+        DutchTrade trade,
+        address bidder,
+        IERC20 sellToken,
+        uint256 soldAmt,
+        IERC20 buyToken,
+        uint256 boughtAmt
+    );
+    DutchTrade private _currentTrade;
+
+    /// Place a bid on an OPEN dutch auction
+    /// @param trade The DutchTrade to bid on
+    /// @param recipient The recipient of the tokens out
+    /// @dev Requires msg.sender has sufficient approval on the tokenIn with router
+    /// @dev Requires msg.sender has sufficient balance on the tokenIn
+    function bid(DutchTrade trade, address recipient) external returns (Bid memory) {
+        Bid memory out = _placeBid(trade, msg.sender);
+        _sendBalanceTo(out.sellToken, recipient);
+        _sendBalanceTo(out.buyToken, recipient);
+        return out;
+    }
+
+    /// @notice Callback for DutchTrade
+    /// @param caller The caller of the callback, should be the router
+    /// @param buyToken The token DutchTrade is expecting to receive
+    /// @param buyAmount The amt the DutchTrade is expecting to receive {qBuyToken}
+    /// @notice Data is not used here
+    function dutchTradeCallback(
+        address caller,
+        address buyToken,
+        uint256 buyAmount,
+        bytes calldata
+    ) external {
+        require(caller == address(this), "Invalid caller");
```

Comment:
I'm not sure I see the point of the `caller` variable. Whichever contract is calling this router contract (DutchTrade or not) can pass in whatever value they want. I think the `msg.sender` check is the only one we need.